### PR TITLE
MSPSDS-928: Filter by team

### PR DESCRIPTION
**Pull request by kicferk1**
The logic behind assignee filters is quite complex now, since we have 3(or more) checkboxes, each of which can be selected or not, and if 'other' is selected but no entity is chosen there is a different bit of logic for who to include. 

I refactored the logic, it passes previous tests, and is easier to follow. I'll check with Eds that this is what we want, and after that I might add more tests for all edge cases. 

There are some minor improvements to add here and there, but that's probably the meat of it.

### CODE REVIEW

File: mspsds-web/app/helpers/investigations_helper.rb [hunk 1]
```diff
@@ -36,36 +36,53 @@ def get_status_filter
...
 
-    if params[:assigned_to_me] == "unchecked" &&
-        params[:assigned_to_someone_else] == "checked" &&
-        params[:assigned_to_someone_else_id].blank?
-      excluded_assignees << current_user.id
-    end
+  def no_boxes_checked
+    no_people_boxes_checked = params[:assigned_to_me] == "unchecked" && params[:assigned_to_someone_else] == "unchecked"
+    no_teams_checked = true
+    teams_with_keys.each { |team| no_teams_checked = no_teams_checked && query_params[team[:key]].blank? }
```
Comment (broder) [inline mspsds-web/app/helpers/investigations_helper.rb]: Would https://apidock.com/ruby/Enumerable/all%3F be helpful here?

File: mspsds-web/app/helpers/investigations_helper.rb [hunk 2]
```diff
@@ -112,4 +130,21 @@ def build_breadcrumb_structure
...
+    current_user.teams.map.with_index do |team, index|
+      {
+        key: "assigned_to_team_#{index}".to_sym,
+        team: team
+      }
+    end
+  end
+
+  def assignee_list_with_team(list, team)
+    list << team.id
```
Comment (broder) [inline mspsds-web/app/helpers/investigations_helper.rb]: Would doing this with https://ruby-doc.org/core-2.2.0/Array.html#method-i-map and https://apidock.com/ruby/Array/concat be neater?

File: mspsds-web/app/helpers/investigations_helper.rb [hunk 3]
```diff
@@ -36,36 +36,53 @@ def get_status_filter
...
+  def assignee_filter_exclusive
+    params[:assigned_to_someone_else] == "checked" && params[:assigned_to_someone_else_id].blank?
+  end
+
+  def compute_excluded_terms
+    excluded_assignees = []
+    teams_with_keys.each do |team_with_key|
+      if query_params[team_with_key[:key]].blank?
+        team = team_with_key[:team]
+        excluded_assignees = assignee_list_with_team(excluded_assignees, team)
```
Comment (broder) [inline mspsds-web/app/helpers/investigations_helper.rb]: Should this be a `concat` or `<<`? (rather than overwriting the excluded_assignees with each team)

File: mspsds-web/app/helpers/investigations_helper.rb [hunk 4]
```diff
@@ -36,36 +36,53 @@ def get_status_filter
...
+      end
     end
+    excluded_assignees << current_user.id if params[:assigned_to_me] == "unchecked"
+    excluded_assignees = excluded_assignees - [current_user.id] if params[:assigned_to_me] == "checked"
```
Comment (broder) [inline mspsds-web/app/helpers/investigations_helper.rb]: Should we do an `if .. else` here? Might make it a bit easier to understand the logic.

File: mspsds-web/app/helpers/investigations_helper.rb [hunk 5]
```diff
@@ -36,36 +36,53 @@ def get_status_filter
...
+    format_assignee_terms(excluded_assignees)
+  end
 
-    if params[:assigned_to_me] == "checked" &&
-        params[:assigned_to_someone_else] == "checked" &&
-        params[:assigned_to_someone_else_id].present?
-      assignees << current_user.id
+  def compute_included_terms
+    assignees = []
+    teams_with_keys.each do |team_with_key|
```
Comment (broder) [inline mspsds-web/app/helpers/investigations_helper.rb]: Could we pull this into a method called `team_assignees` or something like that? Then we could reuse it above.

File: mspsds-web/app/helpers/investigations_helper.rb [hunk 6]
```diff
@@ -36,36 +36,53 @@ def get_status_filter
...
+      if query_params[team_with_key[:key]].present?
+        team = team_with_key[:team]
+        assignees = assignee_list_with_team(assignees, team)
+      end
+    end
+    if params[:assigned_to_someone_else] == "checked"
       assignees << params[:assigned_to_someone_else_id]
+      team = Team.find_by(id: params[:assigned_to_someone_else_id])
+      assignees = assignee_list_with_team(assignees, team) if team.present?
```
Comment (broder) [inline mspsds-web/app/helpers/investigations_helper.rb]: Should this be a `concat` again?

If it was a `concat` and the `assigned_to_someone_else_id` is a team, would the id be added twice?

File: mspsds-web/app/views/investigations/index.html.slim [hunk 7]
```diff
@@ -30,14 +30,18 @@
               .govuk-checkboxes__item
                 = form.check_box :assigned_to_me, { class: "govuk-checkboxes__input" }, "checked", "unchecked"
                 = form.label :assigned_to_me, "Me", class: "govuk-checkboxes__label"
+              - teams_with_keys.each do |team|
```
Comment (broder) [inline mspsds-web/app/views/investigations/index.html.slim]: Could we make this a bit neater with destructuring? http://tony.pitluga.com/2011/08/08/destructuring-with-ruby.html Maybe such that it was something like `teams_with_keys.each do |team, key|`

File: mspsds-web/app/helpers/investigations_helper.rb (hunk 6 above)
Comment (kicferk1) [reply mspsds-web/app/helpers/investigations_helper.rb]: Yet it would.

File: mspsds-web/app/helpers/investigations_helper.rb [hunk 8]
```diff
@@ -36,36 +36,49 @@ def get_status_filter
...
+  end
 
-    if params[:assigned_to_me] == "unchecked" &&
-        params[:assigned_to_someone_else] == "checked" &&
-        params[:assigned_to_someone_else_id].blank?
-      excluded_assignees << current_user.id
-    end
+  def no_boxes_checked
+    no_people_boxes_checked = params[:assigned_to_me] == "unchecked" && params[:assigned_to_someone_else] == "unchecked"
+    no_people_boxes_checked && teams_with_keys.all? { |key, _t, _n| query_params[key].blank? }
```
Comment (broder) [inline mspsds-web/app/helpers/investigations_helper.rb]: I'd maybe keep the `no_team_boxes_checked` variable to make it easier to read

File: mspsds-web/app/helpers/investigations_helper.rb [hunk 9]
```diff
@@ -36,36 +36,49 @@ def get_status_filter
...
-        params[:assigned_to_someone_else] == "checked" &&
-        params[:assigned_to_someone_else_id].present?
-      assignees << current_user.id
+  def compute_included_terms
+    # If 'Me' is not checked, but one of current_users teams is selected, we don't exclude current_user from it
+    assignees = checked_team_assignees
+    if params[:assigned_to_someone_else] == "checked"
       assignees << params[:assigned_to_someone_else_id]
+      team = Team.find_by(id: params[:assigned_to_someone_else_id])
+      assignees.concat(assignee_ids_from_team(team)) if team.present?
```
Comment (broder) [inline mspsds-web/app/helpers/investigations_helper.rb]: Do we want to add the team ID twice?

File: mspsds-web/app/helpers/investigations_helper.rb [hunk 10]
```diff
@@ -36,36 +36,49 @@ def get_status_filter
...
+  end
 
-    assignee_terms = format_assignee_terms(assignees)
-    excluded_assignee_terms = format_assignee_terms(excluded_assignees)
-    { should: assignee_terms, must_not: excluded_assignee_terms }
+  def checked_team_assignees
+    assignees = []
+    teams_with_keys.each do |key, team, _n|
+      if query_params[key].present?
+        team = team
```
Comment (broder) [inline mspsds-web/app/helpers/investigations_helper.rb]: Probably don't need this

File: mspsds-web/test/controllers/team_test.rb [hunk 11]
```diff
@@ -0,0 +1,198 @@
+require "test_helper"
+
+class TeamTest < ActionDispatch::IntegrationTest
+  setup do
+    sign_in_as_user
+    organisations = Organisation.all
+    @teams = Team.all
+    user_groups = [
```
Comment (broder) [inline mspsds-web/test/controllers/team_test.rb]: Should this be in a helper?
